### PR TITLE
Create a hook to manage MinIO connection

### LINKED ISSUE #32872: Unable to connect to minio
### Apache Airflow version

2.6.3

### What happened

[2023-07-26T21:28:19.052+0000] {base_aws.py:160} INFO - No connection ID provided. Fallback on boto3 credential strategy (region_name='\x1b[01mNone\x1b[22m'). See: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html
Out[19]: 
(False,
 "'NoCredentialsError' error occurred while testing connection: Unable to locate credentials")

### What you think should happen instead

Connection works

### How to reproduce

```python
from airflow.models.connection import Connection
conn = Connection.from_json('{"conn_type": "aws", "login": "minio-user", "password": "minio-password", "extra":{"endpoint_url": "http://minio-storage:9000"}}')
conn.test_connection()
````

### Operating System

Linux/docker

### Versions of Apache Airflow Providers

apache-airflow==2.6.3
apache-airflow-providers-amazon==8.2.0

### Deployment

Docker-Compose

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by hussein-awala**
closes: #32872

This PR adds a new hook for MinIO based on S3Hook, just to manage MinIO connection.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by hussein-awala** [comment]
@eladkal @vincbeck 
This draft is just to discuss the idea of creating a new hook for MinIO based on S3Hook, which inherit all the methods except the test_connection (implemented in `AwsGenericHook`), because it is not possible to test it via sts, and IMO implementing a special case for MinIO in AWS hooks is not clean. And if I am not mistaken, the only way to use S3 client with MinIO is providing access and secrets keys with the endpoint url, so we can simplify the UI for MinIO connection by overriding the method  `get_connection_form_widgets`.

WDYT about this and about adding it to AWS provider?

**Comment by vincbeck** [comment]
Overall I am good with this change. I like it. I would just move it to its own file and not be part of the S3 hook file. Also I am not sure about the `test_connection` method. This method is suppose to return the couple `(status, message)`, here nothing is returned.

**Comment by eladkal** [comment]
> I would just move it to its own file and not be part of the S3 hook

I agree.

BTW related issue: https://github.com/apache/airflow/issues/15319
I think this PR solves it as well?

**Comment by hussein-awala** [comment]
@eladkal @vincbeck To use S3 client with MinIO, we should provide the access and the secrets keys, there is no other way for authentication. For that, I simplified the connection UI, and the client creation.

I tested the connection with MinIO docker container, and here is the result:
image
image
image

File: airflow/providers/amazon/aws/hooks/minio.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    def __init__(
+        self,
+        minio_conn_id: str = default_conn_name,
+        transfer_config_args: dict | None = None,
+        extra_args: dict | None = None,
+        *args,
+        **kwargs,
+    ) -> None:
+        kwargs["transfer_config_args"] = transfer_config_args
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/hooks/minio.py]: Why doing so? If you remove this property, the behaviour is the exact same one to me

File: airflow/providers/amazon/aws/hooks/minio.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        kwargs["extra_args"] = extra_args
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/hooks/minio.py]: Same

**Review by vincbeck — APPROVED** [review]

File: airflow/providers/amazon/provider.yaml [hunk 3]
```diff
@@ -499,6 +499,9 @@ hooks:
   - integration-name: Amazon Simple Storage Service (S3)
     python-modules:
       - airflow.providers.amazon.aws.hooks.s3
+  - integration-name: Minio Simple Storage Service (S3)
+    python-modules:
+      - airflow.providers.amazon.aws.hooks.minio
```
Comment (o-nikolas) [inline airflow/providers/amazon/provider.yaml:504]: I'm late to this discussion, why is the AWS provider now responsible for/owns a Minio hook? I know that they support the S3 API, but I'm not sure that warrants us owning and vending that code in the AWS provider package. It sets a bit of a precedent I think.

Comment (potiuk) [reply airflow/providers/amazon/provider.yaml:504]: > I'm late to this discussion, why is the AWS provider now responsible for/owns a Minio hook? I know that they support the S3 API, but I'm not sure that warrants us owning and vending that code in the AWS provider package. It sets a bit of a precedent I think.

Quite agree. It feels out of place here.

Comment (shubham22) [reply airflow/providers/amazon/provider.yaml:504]: 110% agree with Niko here. We should not set this precedence that Amazon provider package can or will own MinIO hook or any other hooks outside of services under AWS. I understand that MinIO is specifically for using S3, but it is not part of AWS portfolio.

Comment (eladkal) [reply airflow/providers/amazon/provider.yaml:504]: I guess it calls for following the protocol of adding new provider
https://github.com/apache/airflow/blob/main/PROVIDERS.rst#accepting-new-community-providers

Comment (potiuk) [reply airflow/providers/amazon/provider.yaml:504]: Pretty much so.

**Review by potiuk — CHANGES_REQUESTED** [review]
Just to prevent accidental merging, since we have pretty universal understanding MinIO should not be part of AWS provider.

Just to add a bit of context/philosophy why. I think it is needed, to not draw bad conclusions ("AWS people do not like it, so it should not be there") for anyone looking - this. is not the case. Also might be a good reminder for those listening.

The providers are not "owned" by the stakeholders, they are part of the community code same as airflow and once we accept them, they are "our" code and we (community) can do what we want - we do not have to ask the stakeholders in them, but we cooperate and collaborate with them - and one of the main reasons of splitting our code to providers was to chunk our code and packages in the pieces that - if possible - if there is a stakeholder that is interested in keep this piece of code maintained, they have it and it's clearly separated (not as clearly yet but we are getting there). This is for example why we decided that transfer operators are in the provider that the stakeholders are most likely to be interested in maintaining. That's why for example GCStoS3 is in amazon and S3 to GCS is in Google. Because generally Amazon is much more interested on people sending data to them, rather than pulling data from them. 

Also that's why we have mixed governance model (not yet excercised but a possibilty that the stakeholder takes responsibility and burden to prepare (but not release) past releases of providers if they see the need for it. Similarly the System Tests that are also connected to particular provider are maintained and run by the stakeholders. All this is part of the same approach -> we collaborate with the stakeholders and make it easy for them to maintain the piece of code they are interested in.

So in this context adding minio to "amazon" provider might seem relevant from "technology" point of view, but when it comes to maintainability properties, it makes no sense to keep it together.

For example - if minio brings a breaking change. Should we make a breaking change for amazon? Or when minio adds dependency that will prevent us from implementing security fixex and move out of end-of-life dpendencies (case of yandex). Should we suspend whole AWS provider because minio dependencies are holding us back?

I think it is pretty straightforward that minio is out-of-place here.

**Comment by eladkal** [comment]
Lets proceed with the procedure of adding new provider as explained in the doc. I think it would not be difficult to get vote passed for this.

**Comment by hussein-awala** [comment]
The main goal of this PR was implementing a quick fix to test the connection with MinIO (which btw used by all AWS users in dev environment) without patching the test connection method in AWSHook.

All of you agree on creating a new provider for MinIO, so I'll close this one and create a meta issue for that.
However, since it will be a separate provider, IMHO we should implement it from scratch using minio client instead of boto3 because it supports all the authentication methods and some other features in MinIO.

**Pull request closed without merging**